sherpa-onnx/csrc/online-rnn-lm.cc

```diff
@@ -0,0 +1,123 @@
+
+    auto out =
+        sess_->Run({}, input_names_ptr_.data(), inputs.data(), inputs.size(),
+                   output_names_ptr_.data(), output_names_ptr_.size());
+
+    std::vector<Ort::Value> next_states;
+    next_states.reserve(2);
+    next_states.push_back(std::move(out[1]));
+    next_states.push_back(std::move(out[2]));
+
+    return {std::move(out[0]), std::move(next_states)};
+  }
+
+  std::vector<Ort::Value> GetInitStates() {
+    constexpr int32_t kBatchSize = 1;
+    std::array<int64_t, 3> h_shape{rnn_num_layers_, kBatchSize,
+                                   rnn_hidden_size_};
+    std::array<int64_t, 3> c_shape{rnn_num_layers_, kBatchSize,
+                                   rnn_hidden_size_};
+    Ort::Value h = Ort::Value::CreateTensor<float>(allocator_, h_shape.data(),
+                                                   h_shape.size());
+    Ort::Value c = Ort::Value::CreateTensor<float>(allocator_, c_shape.data(),
+                                                   c_shape.size());
+    Fill<float>(&h, 0);
+    Fill<float>(&c, 0);
+    std::array<int64_t, 2> x_shape{1, 1};
+    // shape of x and y are same
+    Ort::Value x = Ort::Value::CreateTensor<int64_t>(allocator_, x_shape.data(),
+                                                     x_shape.size());
+    Ort::Value y = Ort::Value::CreateTensor<int64_t>(allocator_, x_shape.data(),
+                                                     x_shape.size());
+    *x.GetTensorMutableData<int64_t>() = sos_id_;
+    *y.GetTensorMutableData<int64_t>() = sos_id_;
+
+    std::vector<Ort::Value> states;
+    states.push_back(std::move(h));
+    states.push_back(std::move(c));
+    auto pair = Rescore(std::move(x), std::move(y), std::move(states));
+    return std::move(pair.second);
+  }
+
+ private:
+  void Init(const OnlineLMConfig &config) {
+    auto buf = ReadFile(config_.model);
+
+    sess_ = std::make_unique<Ort::Session>(env_, buf.data(), buf.size(),
+                                           sess_opts_);
+
+    GetInputNames(sess_.get(), &input_names_, &input_names_ptr_);
+    GetOutputNames(sess_.get(), &output_names_, &output_names_ptr_);
+
+    Ort::ModelMetadata meta_data = sess_->GetModelMetadata();
+    Ort::AllocatorWithDefaultOptions allocator;  // used in the macro below
+    SHERPA_ONNX_READ_META_DATA(rnn_num_layers_, "num_layers");
+    SHERPA_ONNX_READ_META_DATA(rnn_hidden_size_, "hidden_size");
+    SHERPA_ONNX_READ_META_DATA(sos_id_, "sos_id");
+  }
+
+ private:
+  OnlineLMConfig config_;
+  Ort::Env env_;
+  Ort::SessionOptions sess_opts_;
+  Ort::AllocatorWithDefaultOptions allocator_;
+
+  std::unique_ptr<Ort::Session> sess_;
+
+  std::vector<std::string> input_names_;
+  std::vector<const char *> input_names_ptr_;
+
+  std::vector<std::string> output_names_;
+  std::vector<const char *> output_names_ptr_;
+  int32_t rnn_num_layers_ = 2;
+  int32_t rnn_hidden_size_ = 512;
+  int32_t sos_id_ = 1;
```

Comment:
Please add a new member variable
```cpp
std::vector<Ort::Value> init_states_;
```

And in `GetInitStates()` use it like the following:
```
if (!init_states.empty()) {
  return a copy of init_states;
}

// compute init_states
// and then 
// return GetInitStates(); // Yes, do a recursive call
```